test/python/circuit/test_rust_equivalence.py

```diff
@@ -175,3 +190,46 @@ def test_num_params(self):
                 self.assertEqual(
                     len(gate_class.params), standard_gate.num_params, msg=f"{name} not equal"
                 )
+
+    def test_non_default_controls(self):
+        """Test that controlled gates with a non-default ctrl_state
+        are not using the standard rust representation."""
+
+        # tests for CZ, CU1 and CCZ (gates with rust representation)
+        z_gate, z_term = ZGate(), -1
+        u1_gate, u1_term = U1Gate(0.1), 0.99500417 + 0.09983342j
+
+        for gate, term in zip([z_gate, u1_gate], [z_term, u1_term]):
+            with self.subTest(name="2q gate," + gate.name):
+                default_op = np.diag([1, 1, 1, term])
+                non_default_op = np.diag([1, 1, term, 1])
+                state_out_map = {
+                    1: default_op,
+                    "1": default_op,
+                    None: default_op,
+                    0: non_default_op,
+                    "0": non_default_op,
+                }
+                for state, op in state_out_map.items():
+                    circuit = QuantumCircuit(2)
+                    circuit.append(gate.control(1, ctrl_state=state), [0, 1])
+                    np.testing.assert_almost_equal(circuit.data[0].operation.to_matrix(), op)
+
+        with self.subTest(name="3q gate"):
+            default_op = np.diag([1, 1, 1, 1, 1, 1, 1, z_term])
+            non_default_op_0 = np.diag([1, 1, 1, 1, z_term, 1, 1, 1])
+            non_default_op_1 = np.diag([1, 1, 1, 1, 1, z_term, 1, 1])
+            non_default_op_2 = np.diag([1, 1, 1, 1, 1, 1, z_term, 1])
+            state_out_map = {
+                3: default_op,
+                "11": default_op,
+                None: default_op,
+                0: non_default_op_0,
+                1: non_default_op_1,
+                "01": non_default_op_1,
+                "10": non_default_op_2,
+            }
+            for state, op in state_out_map.items():
+                circuit = QuantumCircuit(3)
+                circuit.append(z_gate.control(2, ctrl_state=state), [0, 1, 2])
```

Comment:
Does this actually create a `CCZGate` object? Or does this create a custom `ControlledGate` instance? To test this effectively we'll need to ensure it's actually a `CCZGate` instance because we want both the `._standard_gate` attribute set and also a custom `ctrl_state` .

Reply:
Addressed in [6e7b076](https://github.com/Qiskit/qiskit/pull/12659/commits/6e7b0768e29370aa13c297f9ea9ee15d2f791c42)